src/System.Memory/src/System/SpanHelpers.char.cs

```diff
@@ -0,0 +1,271 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+#if !netstandard
+using Internal.Runtime.CompilerServices;
+#endif
+
+#if !netstandard11
+using System.Numerics;
+#endif
+
+namespace System
+{
+    internal static partial class SpanHelpers
+    {
+        private static readonly Vector<UInt16> s_equalityTester = new Vector<UInt16>(UInt16.MaxValue);
+        private static readonly Vector<UInt16> s_whiteSpace9Mask = new Vector<UInt16>(0x9);
+        private static readonly Vector<UInt16> s_whiteSpace13Mask = new Vector<UInt16>(0xd);
+        private static readonly Vector<UInt16> s_whiteSpace32Mask = new Vector<UInt16>(0x20);
+        private static readonly Vector<UInt16> s_whiteSpace133Mask = new Vector<UInt16>(0x85);
+        private static readonly Vector<UInt16> s_whiteSpace160Mask = new Vector<UInt16>(0xa0);
+        private static readonly Vector<UInt16> s_isLatin1Mask = new Vector<UInt16>(0xFF);
+
+        public static bool IsWhiteSpace(ref char first, int length)
+        {
+            Debug.Assert(length >= 0);
+
+            int i = 0;
+#if !netstandard11
+            if (Vector.IsHardwareAccelerated && length >= 2 * Vector<UInt16>.Count)
+            {
+                while (i < length - Vector<UInt16>.Count)
+                {
+                    Vector<UInt16> value = Unsafe.ReadUnaligned<Vector<UInt16>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, i)));
+                    if (Vector.GreaterThanAny<UInt16>(value, s_isLatin1Mask)) break;
+
+                    if (Vector.GreaterThanAny<UInt16>(value, s_whiteSpace13Mask) ||
+                        Vector.LessThanAny<UInt16>(value, s_whiteSpace9Mask))
```

Comment:
Spaces (0x20) are by far the most common white space. How does this do on list of spaces? Notice that `IsWhiteSpace(char)` is first one that the existing `IsWhitespace` method checks for.

Also, same as for Trim - I would expect this to be called on small Spans. I do not think the vectorization makes sense here.

Reply:
> Spaces (0x20) are by far the most common white space.

Of course. The results I posted above are for a list of ' ' space character. The benefit is coming from doing the IsLatin1 check, vectorized.

> Also, same as for Trim - I would expect this to be called on small Spans. I do not think the vectorization makes sense here.

For Trim/WhiteSpace, there is a 30-40% improvement for span.Length between 16-100 and 60-70% for anything larger (for span.Length == 16, we see the time drop to 60%). Anything smaller than 16 takes a 30-40% perf hit. I would think that is a fair trade-off, especially given how big the gains can get for anything over 15. If the span is smaller, the cost to trim is already relatively low so a regression to that isn't as expensive (unless someone is calling these methods in a loop on small spans).

I am trying to find common uses of equivalent string APIs in our repos to get a sense of the length distribution.
For example, NegotiationResponse.ConnectionID could be >= 16 characters:
https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs#L318

But in that case, we expect it to return false more often than not.

System.Management.SelectQuery can contain large amounts of whitespace:
https://github.com/dotnet/corefx/blob/master/src/System.Management/src/System/Management/ManagementQuery.cs#L660

Similarly, we have NormalizeLocationSubPath - https://github.com/dotnet/corefx/blob/f0ccd9742065ee2aa296448fe3dd38ef6ff77af9/src/System.Configuration.ConfigurationManager/src/System/Configuration/BaseConfigurationRecord.cs#L3102

It is a difficult call. Ideally, we could vectorize the IsLatin1 check and call CharUnicodeInfo.IsWhiteSpace(c) directly, when we move these APIs down to core, which would avoid the redundant IsLatin1 check in Char.IsWhiteSpace().